source/extensions/filters/network/rocketmq_proxy/conn_manager.cc

```diff
@@ -0,0 +1,387 @@
+        group_members_.erase(search);
+      }
+    }
+  }
+
+  RemotingCommandPtr response = std::make_unique<RemotingCommand>(
+      static_cast<int>(ResponseCode::SUCCESS), request->version(), request->opaque());
+  response->markAsResponse();
+  response->remark("Enovy unregister client OK.");
+  sendResponseToDownstream(response);
+  stats_.unregister_.inc();
+}
+
+void ConnectionManager::onError(RemotingCommandPtr& request, absl::string_view error_msg) {
+  Buffer::OwnedImpl buffer;
+  RemotingCommandPtr response = std::make_unique<RemotingCommand>();
+  response->markAsResponse();
+  response->opaque(request->opaque());
+  response->code(static_cast<int>(ResponseCode::SYSTEM_ERROR));
+  response->remark(error_msg);
+  sendResponseToDownstream(response);
+}
+
+void ConnectionManager::onSendMessage(RemotingCommandPtr request) {
+  ENVOY_CONN_LOG(trace, "#onSendMessage, opaque: {}", read_callbacks_->connection(),
+                 request->opaque());
+  createActiveMessage(request).sendRequestToUpstream();
+}
+
+void ConnectionManager::onGetConsumerListByGroup(RemotingCommandPtr request) {
+  auto requestExtHeader =
+      dynamic_cast<GetConsumerListByGroupRequestHeader*>(request->customHeader().get());
+
+  ASSERT(requestExtHeader != nullptr);
+  ASSERT(!requestExtHeader->consumerGroup().empty());
+
+  ENVOY_LOG(trace, "#onGetConsumerListByGroup, consumer group: {}",
+            requestExtHeader->consumerGroup());
+
+  auto search = group_members_.find(requestExtHeader->consumerGroup());
+  GetConsumerListByGroupResponseBody getConsumerListByGroupResponseBody;
+  if (search != group_members_.end()) {
+    std::vector<ConsumerGroupMember>& members = search->second;
+    for (const auto& member : members) {
+      getConsumerListByGroupResponseBody.add(member.clientId());
+    }
+  } else {
+    ENVOY_LOG(warn, "There is no consumer belongs to consumer_group: {}",
+              requestExtHeader->consumerGroup());
+  }
+  ProtobufWkt::Struct body_struct;
+
+  getConsumerListByGroupResponseBody.encode(body_struct);
+
+  RemotingCommandPtr response = std::make_unique<RemotingCommand>(
+      static_cast<int>(ResponseCode::SUCCESS), request->version(), request->opaque());
+  response->markAsResponse();
+  std::string json = MessageUtil::getJsonStringFromMessage(body_struct);
+  response->body().add(json);
+  ENVOY_LOG(trace, "GetConsumerListByGroup respond with body: {}", json);
+
+  sendResponseToDownstream(response);
+  stats_.get_consumer_list_.inc();
+}
+
+void ConnectionManager::onPopMessage(RemotingCommandPtr request) {
+  auto header = dynamic_cast<PopMessageRequestHeader*>(request->customHeader().get());
+  ASSERT(header != nullptr);
+  ENVOY_LOG(trace, "#onPopMessage. Consumer group: {}, topic: {}", header->consumerGroup(),
+            header->topic());
+
+  std::string target_host = app_load_balancer_ptr_->lb(request);
+  if (!target_host.empty()) {
+    createActiveMessage(request).sendRequestToUpstream(target_host);
```

Comment:
sendRequestToUpstream(target_host) ->  createFilterChain(target_host); ->   connection_manager_.config().createRouter(target_host);

```
  RouterPtr createRouter(absl::string_view target_host) override {
    return std::make_unique<RouterImpl>(context_.clusterManager(), target_host);
  }


class RouterImpl {
  private:
      absl::string_view target_host_;
}
```
Finally, target_host_ in RouterImpl refers to the target_host here, but the target_host here is released when it leaves the scope, resulting an undefined behavior in routerimpl.target_host_ refers to invalid memory.

Reply:
Ah, I get your point. This is a bug. target_host_ should be of std::string, not string_view.

Reply:
`string_view` has been fixed already.